node/src/components/consensus/era_supervisor.rs

```diff
@@ -278,124 +262,210 @@ where
+    /// ones, in order, but at most as far back as to the last activation point.
+    pub(super) fn create_new_era<REv: ReactorEventT<I>>(
         &mut self,
-        era_id: EraId,
-        now: Timestamp,
-        validators: BTreeMap<PublicKey, U512>,
-        new_faulty: Vec<PublicKey>,
-        faulty: HashSet<PublicKey>,
-        inactive: HashSet<PublicKey>,
-        seed: u64,
-        start_time: Timestamp,
-        start_height: u64,
-    ) -> Vec<ProtocolOutcome<I, ClContext>> {
+        effect_builder: EffectBuilder<REv>,
+        rng: &mut NodeRng,
+        switch_blocks: &[BlockHeader],
+    ) -> Effects<Event<I>> {
+        let key_block = if let Some(key_block) = switch_blocks.last() {
+            key_block
+        } else {
+            return fatal!(
+                effect_builder,
+                "attempted to create era with no switch blocks; this is a bug",
+            )
+            .ignore();
+        };
+
+        let era_id = key_block.era_id().successor();
+
+        let (report, validators) = match (
+            key_block.era_report(),
+            key_block.next_era_validator_weights(),
+        ) {
+            (Some(report), Some(validators)) => (report, validators.clone()),
+            (_, _) => {
+                return fatal!(
+                    effect_builder,
+                    "attempted to create {} with non-switch block {}; this is a bug",
+                    era_id,
+                    key_block,
+                )
+                .ignore();
+            }
+        };
+
         if self.active_eras.contains_key(&era_id) {
-            panic!("{} already exists", era_id);
+            warn!(era = era_id.value(), "era already exists");
+            return Effects::new();
         }
+        if self.current_era > era_id.saturating_add(2) {
+            warn!(era = era_id.value(), "trying to create obsolete era");
+            return Effects::new();
+        }
+
+        // Compute the seed for the PRNG from the booking block hash and the accumulated seed.
+        let auction_delay = self.protocol_config.auction_delay as usize;
+        let booking_block_hash =
+            if let Some(booking_block) = switch_blocks.iter().rev().nth(auction_delay) {
+                booking_block.hash()
+            } else {
+                // If there's no booking block for the `era_id`
+                // (b/c it would have been from before Genesis, upgrade or emergency restart),
+                // use a "zero" block hash. This should not hurt the security of the leader
+                // selection algorithm.
+                BlockHash::default()
+            };
+        let seed = Self::era_seed(booking_block_hash, key_block.accumulated_seed());
+
+        // The beginning of the new era is marked by the key block.
+        #[allow(clippy::integer_arithmetic)] // Block height should never reach u64::MAX.
+        let start_height = key_block.height() + 1;
```

Comment:
Add to `Block::verify` (or maybe `BlockHeader::verify`?) a check for `height < u64::MAX / 2`.

Reply:
I'm still not sure about this: It's a very different kind of "invalid" than the hashes not matching.

E.g. if we asked for a block by hash, and its hashes don't match (current `verify` would fail), the fetcher should continue asking other peers for the block. But if the hashes do match and only the height is invalid, that means it should fail, and we should drop all blocks that referred to this one by hash.